README.md
# 🔁 @xxmichas/retry

## 💾 Installation

```bash
npm i @xxmichas/retry
```

[npm link](https://www.npmjs.com/package/@xxmichas/retry)

## 📖 Usage

Comments and TypeScript types should be self-explanatory until a proper README is written.

## 📚 TODO

- [ ] README


CHANGELOG.md
# @xxmichas/retry

## 1.1.0

### Minor Changes

- 16c41ec: added cancel to onRetry

## 1.0.0

### Major Changes

- 110dbc6: init


test/index.test.ts
import { retry } from "../src";

it("should return correct value", async () => {
  const value1 = await retry(async () => {
    return {
      some: "value",
    };
  });

  expect(value1).toEqual({
    some: "value",
  });

  let attempt = 0;
  const value2 = await retry(async () => {
    attempt++;

    if (attempt === 1) {
      throw new Error();
    }

    return "value2";
  });

  expect(attempt).toBe(2);
  expect(value2).toEqual("value2");
});

it("should work with synchronous functions", async () => {
  const value = await retry(() => {
    return {
      some: "value",
    };
  });

  expect(value).toEqual({
    some: "value",
  });
});

it("should return after 1st attempt", async () => {
  let attempt = 0;

  await retry(async () => {
    attempt++;
  });

  expect(attempt).toBe(1);
});

it("should return after 2nd attempt", async () => {
  let attempt = 0;

  await retry(async () => {
    attempt++;

    if (attempt === 1) {
      throw new Error();
    }
  });

  expect(attempt).toBe(2);
});

it("shouldn't retry more than 10 times total", async () => {
  let attempt = 0;

  try {
    await retry(
      async () => {
        attempt++;

        throw new Error(`${attempt} err`);
      },
      {
        minTimeout: 5,
        maxTimeout: 5,
      },
    );
  } catch {}

  expect(attempt).toBe(10);
});

it("shouldn't use jitter", async () => {
  let totalTimeout = 0;

  try {
    await retry(
      async () => {
        throw new Error();
      },
      {
        jitter: false,
        minTimeout: 5,
        maxTimeout: 5,
        onRetry: ({ timeout }) => {
          totalTimeout += timeout;
        },
      },
    );
  } catch {}

  expect(totalTimeout).toBe(45);
});

it("should cancel", async () => {
  let attempt = 0;

  try {
    await retry(
      async (cancel) => {
        attempt++;

        if (attempt === 1) {
          cancel(new Error());
        }
      },
      {
        minTimeout: 5,
        maxTimeout: 5,
      },
    );
  } catch {}

  expect(attempt).toBe(1);
});

it("should retry only 3 times", async () => {
  let attempt = 0;

  try {
    await retry(
      async () => {
        attempt++;

        throw new Error();
      },
      {
        attempts: 3,
        minTimeout: 5,
        maxTimeout: 5,
      },
    );
  } catch {}

  expect(attempt).toBe(3);
});

it("should throw an error if first argument is not a function", async () => {
  try {
    // @ts-expect-error
    await retry("invalid");
  } catch (error) {
    expect(error).toBeInstanceOf(TypeError);
    // @ts-ignore
    expect(error.message).toBe("First argument must be a function");
  }
});

it("should pass correct arguments to onRetry", async () => {
  let attempt1 = 0;
  let error1: any;
  let timeout1: number;
  let cancel1: any;

  try {
    await retry(
      async () => {
        attempt1++;

        throw new Error(attempt1.toString());
      },
      {
        minTimeout: 5,
        maxTimeout: 5,
        onRetry: ({ attempt, error, timeout, cancel }) => {
          attempt1 = attempt;
          error1 = error;
          timeout1 = timeout;
          cancel1 = cancel;
        },
      },
    );
  } catch {}

  expect(attempt1).toBe(10);
  expect(error1).toBeInstanceOf(Error);
  expect(error1.message).toBe("9"); // onRetry doesn't run after final attempt
  // @ts-ignore
  expect(timeout1).toBeGreaterThanOrEqual(0);
  // @ts-ignore
  expect(timeout1).toBeLessThanOrEqual(5);
  expect(cancel1).toBeInstanceOf(Function);

  let attempt2 = 0;
  let error2: any;
  let timeout2: number;
  let cancel2: any;

  try {
    await retry(
      async () => {
        attempt2++;

        throw new Error(attempt2.toString());
      },
      {
        minTimeout: 5,
        maxTimeout: 5,
        jitter: false,
        onRetry: ({ attempt, error, timeout, cancel }) => {
          attempt2 = attempt;
          error2 = error;
          timeout2 = timeout;
          cancel2 = cancel;
        },
      },
    );
  } catch {}

  expect(attempt2).toBe(10);
  expect(error2).toBeInstanceOf(Error);
  expect(error2.message).toBe("9"); // onRetry doesn't run after final attempt
  // @ts-ignore
  expect(timeout2).toBe(5);
  expect(cancel2).toBeInstanceOf(Function);
});

it("should cancel in onRetry", async () => {
  let attempt = 0;
  let err: any;

  try {
    await retry(
      async () => {
        attempt++;

        throw new Error();
      },
      {
        minTimeout: 5,
        maxTimeout: 5,
        onRetry: async ({ cancel }) => {
          cancel(new Error("cancel"));
        },
      },
    );
  } catch (error) {
    err = error;
  }

  expect(attempt).toBe(1);
  expect(err).toBeInstanceOf(Error);
  expect(err.message).toBe("cancel");
});

describe("config", () => {
  it("should throw an error if config is invalid", async () => {
    try {
      // @ts-expect-error
      await retry(async () => {}, "invalid");
    } catch (error) {
      expect(error).toBeInstanceOf(TypeError);
      // @ts-ignore
      expect(error.message).toBe("Second argument must be an object");
    }
  });

  it("should throw an error if config.attempts is invalid", async () => {
    try {
      await retry(async () => {}, {
        // @ts-expect-error
        attempts: "invalid",
      });
    } catch (error) {
      expect(error).toBeInstanceOf(TypeError);
      // @ts-ignore
      expect(error.message).toBe(`"config.attempts" must be a number`);
    }

    try {
      await retry(async () => {}, {
        attempts: 0,
      });
    } catch (error) {
      expect(error).toBeInstanceOf(RangeError);
      // @ts-ignore
      expect(error.message).toBe(`"config.attempts" must be greater than 0`);
    }
  });

  it("should throw an error if config.minTimeout is invalid", async () => {
    try {
      await retry(async () => {}, {
        // @ts-expect-error
        minTimeout: "invalid",
      });
    } catch (error) {
      expect(error).toBeInstanceOf(TypeError);
      // @ts-ignore
      expect(error.message).toBe(`"config.minTimeout" must be a number`);
    }

    try {
      await retry(async () => {}, {
        minTimeout: 0,
      });
    } catch (error) {
      expect(error).toBeInstanceOf(RangeError);
      // @ts-ignore
      expect(error.message).toBe(`"config.minTimeout" must be greater than 0`);
    }

    try {
      await retry(async () => {}, {
        minTimeout: 2,
        maxTimeout: 1,
      });
    } catch (error) {
      expect(error).toBeInstanceOf(RangeError);
      // @ts-ignore
      expect(error.message).toBe(`"config.minTimeout" must be less than or equal to "config.maxTimeout"`);
    }
  });

  it("should throw an error if config.maxTimeout is invalid", async () => {
    try {
      await retry(async () => {}, {
        // @ts-expect-error
        maxTimeout: "invalid",
      });
    } catch (error) {
      expect(error).toBeInstanceOf(TypeError);
      // @ts-ignore
      expect(error.message).toBe(`"config.maxTimeout" must be a number`);
    }

    try {
      await retry(async () => {}, {
        maxTimeout: 0,
      });
    } catch (error) {
      expect(error).toBeInstanceOf(RangeError);
      // @ts-ignore
      expect(error.message).toBe(`"config.maxTimeout" must be greater than 0`);
    }
  });

  it("should throw an error if config.jitter is invalid", async () => {
    try {
      await retry(async () => {}, {
        // @ts-expect-error
        jitter: "invalid",
      });
    } catch (error) {
      expect(error).toBeInstanceOf(TypeError);
      // @ts-ignore
      expect(error.message).toBe(`"config.jitter" must be a boolean`);
    }
  });

  it("should throw an error if config.onRetry is invalid", async () => {
    try {
      await retry(async () => {}, {
        // @ts-expect-error
        onRetry: "invalid",
      });
    } catch (error) {
      expect(error).toBeInstanceOf(TypeError);
      // @ts-ignore
      expect(error.message).toBe(`"config.onRetry" must be a function`);
    }
  });
});


src/index.ts
import { checkIfConfigIsValid, FunctionToRetry, RetryConfig, sleep } from "./utils";

export { RetryConfig } from "./utils";

/**
 * Retries a function until it resolves or the maximum amount of attempts is reached.
 *
 * @param fn Function to retry.
 */
export const retry = async <T>(fn: FunctionToRetry<T>, config?: RetryConfig): Promise<T> => {
  if (typeof fn !== "function") {
    throw new TypeError(`First argument must be a function`);
  }

  if (config !== undefined && typeof config !== "object") {
    throw new TypeError(`Second argument must be an object`);
  }

  const options = {
    attempts: config?.attempts ?? 10,
    jitter: config?.jitter ?? true,
    minTimeout: config?.minTimeout ?? 1000,
    maxTimeout: config?.maxTimeout ?? 5000,
    onRetry: config?.onRetry,
  } satisfies RetryConfig;

  checkIfConfigIsValid(options);

  const getTimeout = (attempt: number) => {
    const backoff = Math.min(options.maxTimeout, options.minTimeout * Math.pow(2, attempt - 1));

    if (options.jitter) {
      return Math.random() * backoff;
    }

    return backoff;
  };

  let attempt = 0;
  let isCancelled = false;
  let lastError: any;

  const cancel = (value: any) => {
    isCancelled = true;

    throw value;
  };

  while (attempt < options.attempts) {
    try {
      return await fn(cancel);
    } catch (error) {
      if (isCancelled) {
        throw error;
      }

      attempt++;
      lastError = error;

      if (attempt >= options.attempts) {
        // Last attempt failed
        break;
      }

      const timeout = getTimeout(attempt);

      if (options.onRetry) {
        // If cancel gets called in onRetry, error will be thrown cancelling further execution
        await options.onRetry({ attempt, error, timeout, cancel });
      }

      await sleep(timeout);
    }
  }

  throw lastError;
};


src/utils.ts
export const sleep = async (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

export const checkIfConfigIsValid = (config: Optional<Required<RetryConfig>, "onRetry">) => {
  if (typeof config.attempts !== "number") {
    throw new TypeError(`"config.attempts" must be a number`);
  }

  if (config.attempts < 1) {
    throw new RangeError(`"config.attempts" must be greater than 0`);
  }

  if (typeof config.minTimeout !== "number") {
    throw new TypeError(`"config.minTimeout" must be a number`);
  }

  if (config.minTimeout < 1) {
    throw new RangeError(`"config.minTimeout" must be greater than 0`);
  }

  if (typeof config.maxTimeout !== "number") {
    throw new TypeError(`"config.maxTimeout" must be a number`);
  }

  if (config.maxTimeout < 1) {
    throw new RangeError(`"config.maxTimeout" must be greater than 0`);
  }

  if (config.minTimeout > config.maxTimeout) {
    throw new RangeError(`"config.minTimeout" must be less than or equal to "config.maxTimeout"`);
  }

  if (typeof config.jitter !== "boolean") {
    throw new TypeError(`"config.jitter" must be a boolean`);
  }

  if (config.onRetry !== undefined && typeof config.onRetry !== "function") {
    throw new TypeError(`"config.onRetry" must be a function`);
  }
};

type Optional<T, K extends keyof T> = Pick<Partial<T>, K> & Omit<T, K>;

type CancelFn = (value: any) => void;

export interface FunctionToRetry<T> {
  /**
   * Calling this function rejects the promise with a provided value and stops further retries.
   */
  (cancel: CancelFn): Promise<T> | T;
}

export type RetryConfig = {
  /**
   * Maximum amount of times to retry the operation.
   *
   * Initial attempt counts towards this value.
   *
   * @default 10
   */
  attempts?: number;

  /**
   * Minimum amount of milliseconds to wait before retrying the operation.
   *
   * @default 1000
   */
  minTimeout?: number;

  /**
   * Maximum amount of milliseconds to wait before retrying the operation.
   *
   * @default 5000
   */
  maxTimeout?: number;

  /**
   * Should full jitter be used?
   *
   * @see https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
   *
   * @default true
   */
  jitter?: boolean;

  /**
   * Optional callback to execute before each retry.
   */
  onRetry?: (info: AttemptInfo) => Promise<void> | void;
};

export type AttemptInfo = {
  /**
   * Current attempt.
   */
  attempt: number;
  /**
   * Value thrown by previous attempt.
   */
  error: any;
  /**
   * Number of milliseconds until next retry.
   */
  timeout: number;
  /**
   * Calling this function rejects the promise with a provided value and stops further retries.
   */
  cancel: CancelFn;
};
